Add nvidia-instructlab-bootc and supporting materials

n1hility opened the pull request:
This extends nvidia-bootc:
- Adds the instructlab-nvidia container
- Adds an ilab wrapper script to orchestrate usage of ilab
  + In the near future this will be extended to coordinate the soon to come multi-container mechanism for training but gets us something working now.
- Adds ssh key setup

rhatdan commented:
Why not enhance the training/nvidia directory, which is considered the bootc Containerfile?

rhatdan commented on training/ilab-wrapper/ilab:
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
```
We are going to want this script installed in all bootc containers nvidia, intel and amd?

n1hility replied:
Yes although it needs be changed to be more of a template to support that

rhatdan commented on training/ilab-wrapper/ilab:
```diff
@@ -0,0 +1,23 @@
+		"--security-opt" "label=disable" "--net" "host" \
+		"-v" "$WORKDIR:/instructlab" "--entrypoint" "" \
+		"-e" "HF_HOME=$CONTAINER_CACHE" \
+		"quay.io/ai-lab/instructlab-nvidia:latest")
```
Need a way to make this configurable so that the script can be shared.  Perhaps sed the 

sed -s/VENDOR/${VENDOR}/g /usr/local/bin/ilab

n1hility replied:
Agree I’ll update this to be more of a template

rhatdan commented on training/ilab-wrapper/ilab:
```diff
@@ -0,0 +1,23 @@
+if [[ "$1" = "init" ]]; then
+	shift
+	"${PODMAN_COMMAND[@]}" ilab init \
+		--repository https://github.com/instructlab/taxonomy.git "$@"
```
This is probably not the location that the customer would use, so we need a way to make this flexible.

n1hility replied:
Yeah I should have added something to omit this if already am arg. Only reason it’s there at is because the current default uses SSH git urls and newly provisioned systems are unlikely to have GitHub registered private ssh keys so just keeps it from failing. Ideally we change the default in ilab and we no longer need this.

rhatdan commented on training/ilab-wrapper/ilab:
```diff
@@ -0,0 +1,23 @@
+CONTAINER_CACHE="/instructlab/cache"
+HOST_CACHE="$(pwd)/cache"
+WORKDIR="$(pwd)"
+mkdir -p "$HOST_CACHE"
```
nit ${HOST_CACHE}

rhatdan commented on training/ilab-wrapper/ilab:
```diff
@@ -0,0 +1,23 @@
+PODMAN_COMMAND=("podman" "run" "--rm" "-it" "--device" "nvidia.com/gpu=all" \
+		"--security-opt" "label=disable" "--net" "host" \
+		"-v" "$WORKDIR:/instructlab" "--entrypoint" "" \
+		"-e" "HF_HOME=$CONTAINER_CACHE" \
```
nit ${CONTAINER_CACHE}

n1hility commented:
> Why not enhance the training/nvidia directory, which is considered the bootc Containerfile?

Ah I thought there was a pattern of one sibling per image for the 6 images. So you prefer a subdir? Or would rather it all be in the nvidia directory just an additional make task?

rhatdan commented:
Yes although I am not sure why we need alternatives. to the nvidia Containerfile?

n1hility committed:
Add nvidia-instructlab-bootc and supporting materials

This extends nvidia-bootc:
  - Adds the instructlab
  - Uses a save/load approach instead of pull to avoid registry  issues
  - Adds an ilab wrapper script to orchestrate usage of ilab
    + In the near future this will be extended to coordinate the soon
      to come multi-container mechanism for training but gets us
      something working now.
  - Adds ssh key setup

Signed-off-by: Jason T. Greene <jason.greene@redhat.com>

n1hility commented:
Refactored this. In addition to discussed changes I needed to switch to a save/load approach due to a corrupt instructlab-nvidia image on quay. Unfortunately this means the multi-container work ended up getting delayed.

Note that to build this you need to make nvidia in instruct-lab first so that its in the local container storage.

PTAL @rhatdan

rhatdan commented on training/nvidia-bootc/Containerfile:
```diff
@@ -123,16 +123,38 @@ RUN if [ "${TARGET_ARCH}" == "" ]; then \
     && echo "blacklist nouveau" > /etc/modprobe.d/blacklist_nouveau.conf
 
-ARG INSTRUCTLAB_IMAGE="quay.io/ai-lab/instructlab-nvidia:latest"
+
```
I think we should separate this out into a separate Containerfile, If we are going to make the images public by default, we don't want to embed the builders ssh keys into the container.  

Perhaps 
```
cat Containerfile.sshpubkey
args from="quay.io/ai-lab/HARDWARE-bootc"
FROM ${from}

ARG SSHPUBKEY

# The --build-arg "SSHPUBKEY=$(cat ~/.ssh/id_rsa.pub)" option inserts your
# public key into the image, allowing root access via ssh.
RUN set -eu; mkdir -p /usr/ssh && \
    echo 'AuthorizedKeysFile /usr/ssh/%u.keys .ssh/authorized_keys .ssh/authorized_keys2' >> /etc/ssh/sshd_config.d/30-auth-system.conf && \
    echo ${SSHPUBKEY} > /usr/ssh/root.keys && chmod 0600 /usr/ssh/root.keys
```

n1hility replied:
Good point will pull this out

rhatdan commented on training/nvidia-bootc/Containerfile:
```diff
@@ -123,16 +123,38 @@ RUN if [ "${TARGET_ARCH}" == "" ]; then \
+    sed -i -e '/additionalimage.*/a "/usr/lib/containers/storage",' \
         /etc/containers/storage.conf
 
+ARG WRAPPER
```
Can you default this.
Something like:
ARG WRAPPER=ilab.script

rhatdan commented on training/nvidia-bootc/Containerfile:
```diff
@@ -123,16 +123,38 @@ RUN if [ "${TARGET_ARCH}" == "" ]; then \
+ARG WRAPPER
+COPY ${WRAPPER} /usr/local/bin/ilab
+
+ARG INSTRUCTLAB_IMAGE
```
Default it.

rhatdan commented on training/nvidia-bootc/Containerfile:
```diff
@@ -123,16 +123,38 @@ RUN if [ "${TARGET_ARCH}" == "" ]; then \
+COPY ${WRAPPER} /usr/local/bin/ilab
+
+ARG INSTRUCTLAB_IMAGE
+ARG INSTRUCTLAB_IMAGE_ID
```
Ditto

rhatdan commented on training/nvidia-bootc/Containerfile:
```diff
@@ -123,16 +123,38 @@ RUN if [ "${TARGET_ARCH}" == "" ]; then \
+
+ARG INSTRUCTLAB_IMAGE
+ARG INSTRUCTLAB_IMAGE_ID
+ARG INSTRUCTLAB_IMAGE_SHA
```
Ditto

rhatdan commented:
Going to merge and then we can iterate.
